text/0000-jsbox.md

```diff
@@ -0,0 +1,341 @@
+
+```rust
+fn create_pool(mut cx: FunctionContext) -> JsResult<JsBox<RefCell<Pool>>> {
+    let pool = RefCell::new(Pool::new());
+
+    cx.boxed(pool)
+}
+
+fn set_pool_size(mut cx: FunctionContext) -> JsResult<JsUndefined> {
+    let size = cx.argument::<JsNumber>(1)?.value() as u32;
+    let mut pool = cx.argument::<JsPool<RefCell<Pool>>>(0)?;
+
+    pool.borrow_mut().set_size(size);
+
+    Ok(cx.undefined())
+}
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+### `neon::types::JsBox`
+
+```rust
+impl<T: Send + 'static> JsValue for JsBox<T> {}
+
+impl<T: Send + 'static> Managed for JsBox<T> {}
+
+impl<T: Send + 'static> ValueInternal for JsBox<> {}
+
+impl<T> JsBox where T: Send + 'static {
+    /// This should only fail if the VM is in a throwing state or is
+    /// in the process of stopping.
+    pub fn new<'a, C: Context<'a>>(
+        cx: &mut C,
+        v: T,
+    ) -> NeonResult<JsBox>;
+}
+```
+
+`JsBox` can be passed back and forth between Javascript and Rust like any other Js value.
+
+
+#### `std::ops::Deref`
+
+`JsBox<T>` acts as a smart-pointer and can be coerced to a `&T` by auto-deref.
+
+```rust
+impl<'a, T: Send + 'static> Deref for JsBox<T> {
+    type Target = T;
+
+    fn deref(&self) -> &Self::Target;
+}
+```
+
+It is not possible to statically prove that two `JsBox<T>` do not alias the same pointer. For example, if a user passes the same value as two arguments to a function. Therefore, it is only safe for `JsBox<T>` to implement `Deref` and not `DerefMut`.
+
+#### `Send + 'static`
+
+`JsBox` are passed across threads and therefore must be `Send`. It is *not* necessary for `JsBox` to be `Sync` because they may only be borrowed on the main thread.
+
+### `trait Context`
+
+A convenience method is provided on the `Context` trait for boxing types. The term `boxed` was chosen because `box` is a keyword.
+
+```rust
+trait Context<'a> {
+    pub fn boxed<U: Send + 'static>(
+        &mut self,
+        v: U,
+    ) -> Handle<'a, JsBox<U>> {
+        JsBox::new(self, v)
+    }
+}
+```
+
+#### Implementation Notes
+
+`JsBox` combines the properties of several N-API and Rust features to produce a powerful API for safely passing Rust structs across the FFI boundary.
+
+##### N-API
+
+* `napi_create_external()`. This API allocates a JavaScript value with external data attached to it.
+* `napi_get_value_external()`. This API retrieves the external data pointer that was previously passed `to napi_create_external()`.
+* `napi_typeof()`. Returns `napi_external` if the type is an external type.
+
+*Note*: Passing an external created by another library, potentially even another neon library, is _undefined behavior_.
+
+Progress is being made to add a [tagging feature](https://github.com/nodejs/node/pull/28237) to more safely unwrap externals. It should be incorporated into future designs; however, it is not included in this proposal.
```

Comment:
This is a little spare on details. Can you explain what extra level of guarantee the tagging feature will provide for the `JsBox` API? What guarantee will we be lacking until we have it available? Or will it only help clean up the implementation?

Reply:
This is directly related to the note above. I'll reformat to make it clearer that they are related.